File: leetcode/206-reverse-linked-list.cpp
/*
Reverse a singly linked list.

*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return head;
        }
        auto p1 = head;
        auto p2 = head->next;
        head->next = nullptr;
        while (p2 != nullptr) {
            auto tmp = p2->next;
            p2->next = p1;
            p1 = p2;
            p2 = tmp;
        }

        return p1;
    }
};

File: leetcode/206-reverse-linked-list.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {ListNode}
def reverse_list(head)
  return nil if head.nil?
  p1 = head
  p2 = head.next
  p1.next = nil
  while not p2.nil?
    nxt = p2.next
    p2.next = p1
    p1 = p2
    p2 = nxt
  end
  p1
end
